tests/unit/s2n_ktls_io_test.c

```diff
@@ -74,5 +102,206 @@ int main(int argc, char **argv)
         };
     };
 
+    /* Test s2n_ktls_sendmsg */
+    {
+        /* Safety */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_test_ktls_io_stuffer_pair io_pair = { 0 },
+                    s2n_ktls_io_stuffer_pair_free);
+            EXPECT_OK(s2n_test_init_ktls_io_stuffer(server, client, &io_pair));
+
+            struct iovec msg_iov_valid = { .iov_base = test_data, .iov_len = S2N_TEST_TO_SEND };
+            s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+            size_t bytes_written = 0;
+
+            EXPECT_ERROR_WITH_ERRNO(
+                    s2n_ktls_sendmsg(server, test_record_type, NULL, 1, &blocked, &bytes_written),
+                    S2N_ERR_NULL);
+            EXPECT_ERROR_WITH_ERRNO(
+                    s2n_ktls_sendmsg(server, test_record_type, &msg_iov_valid, 1, NULL, &bytes_written),
+                    S2N_ERR_NULL);
+            EXPECT_ERROR_WITH_ERRNO(
+                    s2n_ktls_sendmsg(server, test_record_type, &msg_iov_valid, 1, &blocked, NULL),
+                    S2N_ERR_NULL);
+
+            struct iovec msg_iov_null_data = { .iov_base = NULL, .iov_len = S2N_TEST_TO_SEND };
+            EXPECT_ERROR_WITH_ERRNO(
+                    s2n_ktls_sendmsg(server, test_record_type, &msg_iov_null_data, 1, &blocked, &bytes_written),
+                    S2N_ERR_INVALID_ARGUMENT);
+        };
+
+        /* Happy case: msg_iovlen = 1 */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_test_ktls_io_stuffer_pair io_pair = { 0 },
+                    s2n_ktls_io_stuffer_pair_free);
```

Comment:
Are you creating `client` just because s2n_test_init_ktls_io_stuffer takes two connections? Might be worth a variant that only takes one connection

Reply:
Not sure I like this version. There are alot more init functions, you now have to think about the server only init. Also we only save a single client init in the tests which really is not a huge deal since it made all the tests look the same so was actually kinda ok.

https://github.com/aws/s2n-tls/commit/72fa856c4fb1c6c0713be147947abe49d3cedf90

Reply:
Actually I was thinking in terms of `io_pair`. Here is a simpler version that I actually like: https://github.com/aws/s2n-tls/pull/4147/commits/1f3fa42f90248450eac606378cf86d4a463058db